File: LeetCode/1889. Minimum Space Wasted From Packaging/solution.py
# class Solution:
#     def minWastedSpace(self, A, boxes):
#         A.sort()
#         res = float('inf')
#         for B in boxes:
#             B.sort()
#             if B[-1] < A[-1]: continue
#             cur = i = 0
#             for b in B:
#                 j = bisect.bisect(A, b, i)
#                 cur += b * (j - i)
#                 i = j
#             res = min(res, cur)
#         return (res - sum(A)) % (10**9 + 7) if res < float('inf') else -1        

class Solution:
    def minWastedSpace(self, packages: List[int], boxes: List[List[int]]) -> int:
        MOD = 10**9+7
        maxn = 10**5+5
        presums = [0]*maxn
        precnt = [0]*maxn
        n = len(packages)
        heapq.heapify(packages)
        for i in range(1,maxn):
            presums[i] = presums[i-1]
            precnt[i] = precnt[i-1]
            while packages and packages[0] == i:
                presums[i] += packages[0]
                heapq.heappop(packages)
                precnt[i] += 1  
        res = float('inf')
        for box in boxes:
            cur, sums, cnt = 0, 0, 0 
            for b in sorted(box):
                cur += b*(precnt[b]-cnt)-(presums[b]-sums)
                sums = presums[b]
                cnt = precnt[b]
            if cnt < n:
                continue
            res = min(res, cur)
        return -1 if res == float('inf') else (res%MOD)

File: LeetCode/1889. Minimum Space Wasted From Packaging/solution.cpp
const int inf = 0x3f3f3f3f;
const int maxn = 1e5+10;
const int MOD = 1e9+7;
using ll = long long;

int cnt[maxn];


class Solution {
public:
    int minWastedSpace(vector<int>& packages, vector<vector<int>>& boxes) {
        memset(cnt, 0, maxn*sizeof(int));
        ll tot = 0;
        int n = packages.size();
        for (auto p : packages) cnt[p]++, tot += p;
        for (int i = 1; i < maxn; i++) cnt[i] += cnt[i-1];
        ll res = LONG_MAX;
        for (auto &box : boxes){
            sort(box.begin(), box.end());
            ll sums = 0, pre = 0, cur = 0;
            for (auto b : box){
                cur = cnt[b];
                sums += (cur-pre)*b;
                pre = cur;
                if (cur == n)
                    break;
            }
            if (cur < n)
                continue;
            res = min(res, sums-tot);
        }
        return res == LONG_MAX ? -1 : res%MOD;
    }
};


class Solution {
public:
    int minWastedSpace(vector<int>& A, vector<vector<int>>& boxes) {
        sort(A.begin(), A.end());
        long res = LONG_MAX, mod = 1e9 + 7, sumA = 0;
        // cout << res << endl;
        for (int a : A)
            sumA += a;
        for (auto& B : boxes) {
            sort(B.begin(), B.end());
            if (B[B.size() - 1] < A[A.size() - 1]) continue;
            long cur = 0, i = 0, j;
            for (int b : B) {
                j = upper_bound(A.begin(), A.end(), b) - A.begin();
                cur += b * (j - i);
                i = j;
            }
            res = min(res, cur);
        }
        return res < LONG_MAX ? (res - sumA) % mod : -1;
    }
};



// using ll = long long;

// class Solution {
// public:
//     int minWastedSpace(vector<int>& packages, vector<vector<int>>& boxes) {
//         const int MOD = 1e9+7;
//         const ll inf = 1e10+1;
//         sort(packages.begin(), packages.end());
//         vector<ll> presums = {0};
//         for (auto p : packages)
//             presums.push_back(presums.back()+p);
//         for (auto &box : boxes)
//             sort(box.begin(), box.end());
//         int n = boxes.size(), m = packages.size();
//         vector<ll> ans(n,0);
//         for (int i = 0; i < n; i++){
//             int pre = 0;
//             for (int j = 0; j < boxes[i].size(); j++){
//                 int cur = upper_bound(packages.begin(), packages.end(), boxes[i][j])-packages.begin();
//                 ans[i] += static_cast<ll>(boxes[i][j])*(cur-pre)-(presums[cur]-presums[pre]);
//                 pre = cur;
//             }
//             ans[i] = pre < m ? inf : ans[i];
//         }
//         ll res = *min_element(ans.begin(), ans.end());
//         return res < inf ? res%MOD : -1;
//     }
// };